Dependencies/xr/Source/OpenXR/SceneUnderstanding.cpp

```diff
@@ -0,0 +1,497 @@
+
+using namespace xr;
+
+constexpr SceneUnderstanding::DetectionBoundaryType DEFAULT_BOUNDARY_TYPE = SceneUnderstanding::DetectionBoundaryType::Sphere;
+constexpr float DEFAULT_SPHERE_RADIUS = 5.f;
+constexpr double DEFAULT_UPDATE_INTERVAL_IN_SECONDS = 2;
+constexpr double NANOSECONDS_IN_SECOND = 1000000000;
+
+inline xr::Pose XrPoseToBabylonPose(XrPosef pose)
+{
+    // Right handed to left handed conversions are handled in the Babylon.js layer
+    return xr::Pose{
+        {
+            pose.position.x,
+            pose.position.y,
+            pose.position.z,
+        },
+        {
+            pose.orientation.x,
+            pose.orientation.y,
+            pose.orientation.z,
+            pose.orientation.w
+        }};
+}
+
+SceneUnderstanding::SceneUnderstanding::InitOptions::InitOptions(
+    const XrSession& session,
+    const XrSupportedExtensions& extensions)
+    : Session(session)
+    , Extensions(extensions)
+    , DetectionBoundaryType(DEFAULT_BOUNDARY_TYPE)
+    , SphereRadius(DEFAULT_SPHERE_RADIUS)
+    , BoxDimensions(XrVector3f{ 0, 0, 0 })
+    , UpdateIntervalInSeconds(DEFAULT_UPDATE_INTERVAL_IN_SECONDS) {}
+
+SceneUnderstanding::SceneUnderstanding::InitOptions::InitOptions(
+    const XrSession& session,
+    const XrSupportedExtensions& extensions,
+    const SceneUnderstanding::DetectionBoundaryType detectionBoundaryType,
+    const float sphereRadius,
+    const XrVector3f boxDimensions,
+    const double updateIntervalInSeconds)
+    : Session(session)
+    , Extensions(extensions)
+    , DetectionBoundaryType(detectionBoundaryType)
+    , SphereRadius(sphereRadius)
+    , BoxDimensions(boxDimensions)
+    , UpdateIntervalInSeconds(updateIntervalInSeconds) {}
+
+SceneUnderstanding::SceneUnderstanding::UpdateFrameArgs::UpdateFrameArgs(
+    const XrSpace& sceneSpace,
+    const XrSupportedExtensions& extensions,
+    const XrTime displayTime,
+    std::vector<System::Session::Frame::Plane>& planes,
+    std::vector<System::Session::Frame::Plane::Identifier>& updatedPlanes,
+    std::vector<System::Session::Frame::Plane::Identifier>& removedPlanes)
+    : SceneSpace(sceneSpace)
+    , Extensions(extensions)
+    , DisplayTime(displayTime)
+    , Planes(planes)
+    , UpdatedPlanes(updatedPlanes)
+    , RemovedPlanes(removedPlanes)
+{
+}
+
+SceneUnderstanding::SceneUnderstanding::UpdateSceneObjectsArgs::UpdateSceneObjectsArgs(
+    const XrSpace& sceneSpace,
+    const XrSupportedExtensions& extensions,
+    const XrTime displayTime,
+    std::map<XrSceneObjectKeyMSFT, std::shared_ptr<SceneObject>>& sceneObjects,
+    std::vector<XrSceneObjectKeyMSFT>& updatedObjects,
+    std::vector<XrSceneObjectKeyMSFT>& removedObjects)
+    : SceneSpace(sceneSpace)
+    , Extensions(extensions)
+    , DisplayTime(displayTime)
+    , SceneObjects(sceneObjects)
+    , UpdatedObjects(updatedObjects)
+    , RemovedObjects(removedObjects)
+{
+}
+
+struct SceneUnderstanding::SceneUnderstanding::Impl
+{
+public:
+    Impl() {}

```

Comment:
nit: If this is empty, can we omit it? Not sure what the convention in this project is.

Also add a break between this and the next function.